src/HookCallbackRule.php

```diff
@@ -0,0 +1,218 @@
+    {
+        $acceptedArgs = 1;
+
+        if (isset($acceptedArgsParam)) {
+            $acceptedArgs = null;
+            $argumentType = $this->currentScope->getType($acceptedArgsParam->value);
+
+            if ($argumentType instanceof ConstantIntegerType) {
+                $acceptedArgs = $argumentType->getValue();
+            }
+        }
+
+        return $acceptedArgs;
+    }
+
+    protected function validateParamCount(ParametersAcceptor $callbackAcceptor, ?Arg $acceptedArgsParam): void
+    {
+        $acceptedArgs = $this->getAcceptedArgs($acceptedArgsParam);
+
+        if ($acceptedArgs === null) {
+            return;
+        }
+
+        $allParameters = $callbackAcceptor->getParameters();
+        $requiredParameters = array_filter(
+            $allParameters,
+            static function (\PHPStan\Reflection\ParameterReflection $parameter): bool {
+                return ! $parameter->isOptional();
+            }
+        );
+        $maxArgs = count($allParameters);
+        $minArgs = count($requiredParameters);
+
+        if (($acceptedArgs >= $minArgs) && ($acceptedArgs <= $maxArgs)) {
+            return;
+        }
+
+        if (($acceptedArgs >= $minArgs) && $callbackAcceptor->isVariadic()) {
+            return;
+        }
+
+        if ($minArgs === 0 && $acceptedArgs === 1) {
+            return;
+        }
+
+        throw new \SzepeViktor\PHPStan\WordPress\HookCallbackException(
+            self::buildParameterCountMessage(
+                $minArgs,
+                $maxArgs,
+                $acceptedArgs,
+                $callbackAcceptor
+            )
+        );
+    }
+
+    protected static function buildParameterCountMessage(int $minArgs, int $maxArgs, int $acceptedArgs, ParametersAcceptor $callbackAcceptor): string
+    {
+        $expectedParametersMessage = $minArgs;
+
+        if ($maxArgs !== $minArgs) {
+            $expectedParametersMessage = sprintf(
+                '%1$d-%2$d',
+                $minArgs,
+                $maxArgs
+            );
+        }
+
+        $message = ($expectedParametersMessage === 1)
+            ? 'Callback expects %1$d parameter, $accepted_args is set to %2$d.'
+            : 'Callback expects %1$s parameters, $accepted_args is set to %2$d.';
+
+        if ($callbackAcceptor->isVariadic()) {
+            $message = ($minArgs === 1)
+                ? 'Callback expects at least %1$d parameter, $accepted_args is set to %2$d.'
+                : 'Callback expects at least %1$s parameters, $accepted_args is set to %2$d.';
+        }
+
+        return sprintf(
+            $message,
+            $expectedParametersMessage,
+            $acceptedArgs
+        );
+    }
+
+    protected function validateActionReturnType(ParametersAcceptor $callbackAcceptor): void
+    {
+        $acceptedType = $callbackAcceptor->getReturnType();
+        $accepted = $this->ruleLevelHelper->accepts(
+            new VoidType(),
```

Comment:
I never even used it myself yet, but does it make sense maybe to also add tests for  the `never` return type? Not sure what we'd expect, e.g. if PHPStan or this rule is going to complain that the callable never returns / always exits 🤔

Reply:
I'll take a look!

Reply:
Ah I've already got tests for this! https://github.com/szepeviktor/phpstan-wordpress/pull/107/files#diff-56e889cfab3a17dd1d1fbc7b198fcf3130ca973b4a7fd3342c696d29df13a96dR199-R213.

Unfortunately it is acceptable for both filters and actions to exit because WP YOLO. I'm not sure how `VoidType` relates to `NeverType` but this is working as expected and the tests pass, so I think we're all good here.

Reply:
oh, ok. I missed those tests